### PR TITLE
20_2 TreeList Column Reordering: rewrite

### PR DESCRIPTION
**Pull request by pavel-2020**
[Trello](https://trello.com/c/Bi9lS4Ee/1654-treelist-demos-column-customization-4-demos)
CPs: https://github.com/DevExpress/devextreme-demos/pull/915, https://github.com/DevExpress/devextreme-demos/pull/916

### CODE REVIEW

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 1]
```diff
@@ -1 +1,14 @@
-At&nbsp;first launch, the order of&nbsp;columns in&nbsp;TreeList is&nbsp;the same as&nbsp;the order of&nbsp;objects in&nbsp;the **columns** array. At&nbsp;runtime, a&nbsp;user may be&nbsp;able to&nbsp;reorder columns using drag and drop. To&nbsp;provide the user with this capability, set the **allowColumnReordering** property to _true_.
\ No newline at end of file
+The initial order of columns mirrors the order of fields in the first object from the [dataSource](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#dataSource). You can change that order programmatically and allow users to reorder columns in the TreeList's interface.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/Reordering/description.md]: The first sentence is not true, see the screenshot. In my opinion, the whole concept is misleading. Are there many cases when users create a TreeList without the **columns** property?

image

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 2]
```diff
@@ -1 +1,14 @@
...
+
+**Properties to reorder columns programmatically:**
+
+* [columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/)
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
* **[columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/)**           
```

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 3]
```diff
@@ -1 +1,14 @@
-At&nbsp;first launch, the order of&nbsp;columns in&nbsp;TreeList is&nbsp;the same as&nbsp;the order of&nbsp;objects in&nbsp;the **columns** array. At&nbsp;runtime, a&nbsp;user may be&nbsp;able to&nbsp;reorder columns using drag and drop. To&nbsp;provide the user with this capability, set the **allowColumnReordering** property to _true_.
\ No newline at end of file
+The initial order of columns mirrors the order of fields in the first object from the [dataSource](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#dataSource). You can change that order programmatically and allow users to reorder columns in the TreeList's interface.
+
+**Properties to reorder columns programmatically:**
+
+* [columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/)           
+Configures columns. Arrange column names and objects in the desired order.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
Configures columns. Arrange column names or data objects in the desired order.
```

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 4]
```diff
@@ -1 +1,14 @@
...
+
+* **columns[]**.[visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex)          
+Specifies the position of the column regarding other columns. For details, see the [Column Reordering](https://js.devexpress.com/Documentation/Guide/UI_Components/TreeList/Columns/Column_Reordering/) article.
+
+**Property that enables users to reorder columns:**
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/Reordering/description.md]: I'm not sure that a list with only one item is a good idea. I think that it's better to transform the list to a regular paragraph or move the **allowReordering** description to a separate list item.

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 5]
```diff
@@ -1 +1,14 @@
...
\ No newline at end of file
+The initial order of columns mirrors the order of fields in the first object from the [dataSource](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#dataSource). You can change that order programmatically and allow users to reorder columns in the TreeList's interface.
+
+**Properties to reorder columns programmatically:**
+
+* [columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/)           
+Configures columns. Arrange column names and objects in the desired order.
+
+* **columns[]**.[visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex)          
+Specifies the position of the column regarding other columns. For details, see the [Column Reordering](https://js.devexpress.com/Documentation/Guide/UI_Components/TreeList/Columns/Column_Reordering/) article.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
Specifies the position of the column regarding other columns. For details, see the [Column Reordering](/Documentation/Guide/UI_Components/TreeList/Columns/Column_Reordering/) article.
```

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 6]
```diff
@@ -1 +1,7 @@
-At&nbsp;first launch, the order of&nbsp;columns in&nbsp;TreeList is&nbsp;the same as&nbsp;the order of&nbsp;objects in&nbsp;the **columns** array. At&nbsp;runtime, a&nbsp;user may be&nbsp;able to&nbsp;reorder columns using drag and drop. To&nbsp;provide the user with this capability, set the **allowColumnReordering** property to _true_.
\ No newline at end of file
+The structure of [dataSource](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#dataSource) objects defines the order of TreeList columns. You can change the order programmatically and allow users to reorder columns in the UI.
+
+### Reorder columns programmatically
+To reorder columns in code, arrange objects and column names as strings in the desired order inside the [columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array. Additionally, you can use the **columns[]**.[visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property to specify  the position of the column regarding other columns. For details, see the [Column Reordering](/Documentation/Guide/UI_Components/TreeList/Columns/Column_Reordering/) article.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
To reorder columns in code, arrange objects and column names as strings in the desired order inside the [columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array. Additionally, you can use the **columns[]**.[visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property to specify the position of the column regarding other columns. For details, see the [Column Reordering](/Documentation/Guide/UI_Components/TreeList/Columns/Column_Reordering/) article.
```

**Review by AlexJameson — DISMISSED** [review]

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 7]
```diff
@@ -1 +1,7 @@
-At&nbsp;first launch, the order of&nbsp;columns in&nbsp;TreeList is&nbsp;the same as&nbsp;the order of&nbsp;objects in&nbsp;the **columns** array. At&nbsp;runtime, a&nbsp;user may be&nbsp;able to&nbsp;reorder columns using drag and drop. To&nbsp;provide the user with this capability, set the **allowColumnReordering** property to _true_.
\ No newline at end of file
+The structure of [dataSource](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#dataSource) objects defines the order of TreeList columns. You can change the order programmatically and allow users to reorder columns in the UI.
```
Comment (RomanTsukanov) [inline JSDemos/Demos/TreeList/Reordering/description.md]: > The structure of [dataSource](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#dataSource) objects defines the order of TreeList columns.

First the old ("the order of columns"), then the new ("is defined by the structure of data objects").

Also, there's little information to divide the description into subheadings, so I moved information from the following paragraph here.

```suggestion
The order of TreeList columns is the same as the order of fields in data objects. You can use the [columns](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array to specify a different order. To reorder a column at runtime, change its [visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property.
```

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 8]
```diff
@@ -1 +1,7 @@
...
+
+### Reorder columns programmatically
+To reorder columns in code, arrange objects and column names as strings in the desired order inside the [columns[]](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array. Additionally, you can use the **columns[]**.[visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property to specify the position of the column regarding other columns. For details, see the [Column Reordering](/Documentation/Guide/UI_Components/TreeList/Columns/Column_Reordering/) article.
```
Comment (RomanTsukanov) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
```

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 9]
```diff
@@ -1 +1,7 @@
...
+
+### Allow users to reorder columns in the UI
+To enable column reorder in the UI, set [allowColumnReordering](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#allowColumnReordering) to **true**. If some column should never be reordered, set its [allowReordering](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowReordering) property to **false**.
```
Comment (RomanTsukanov) [inline JSDemos/Demos/TreeList/Reordering/description.md]: I added how users reorder columns and change this bit:

> If some column should never be reordered

to make clear what exactly happens when you disable a column's `allowReordering`.

```suggestion
Users can drag and drop column headers to reorder columns. To enable this feature, set the [allowColumnReordering](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/#allowColumnReordering) property to **true**. If you do not want users to drag a specific column, disable its [allowReordering](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowReordering) property. In this demo, users cannot drag the Full Name column.
```

**Review by RomanTsukanov — CHANGES_REQUESTED** [review]

**Review by RomanTsukanov — DISMISSED** [review]

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 10]
```diff
@@ -1 +1,3 @@
-At&nbsp;first launch, the order of&nbsp;columns in&nbsp;TreeList is&nbsp;the same as&nbsp;the order of&nbsp;objects in&nbsp;the **columns** array. At&nbsp;runtime, a&nbsp;user may be&nbsp;able to&nbsp;reorder columns using drag and drop. To&nbsp;provide the user with this capability, set the **allowColumnReordering** property to _true_.
\ No newline at end of file
+The order of TreeList columns is the same as the order of fields in data objects. You can use the [columns](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array to specify a different order. To reorder a column at runtime, change its [visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property.
```
Comment (dmitry-eliseev-devexpress) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
The order of TreeList columns follows the order of fields in data objects. You can use the [columns](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array to specify a different order. To reorder a column at runtime, change its [visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property.
```

**Review by dmitry-eliseev-devexpress — DISMISSED** [review]

File: JSDemos/Demos/TreeList/Reordering/description.md [hunk 11]
```diff
@@ -1 +1,3 @@
-At&nbsp;first launch, the order of&nbsp;columns in&nbsp;TreeList is&nbsp;the same as&nbsp;the order of&nbsp;objects in&nbsp;the **columns** array. At&nbsp;runtime, a&nbsp;user may be&nbsp;able to&nbsp;reorder columns using drag and drop. To&nbsp;provide the user with this capability, set the **allowColumnReordering** property to _true_.
\ No newline at end of file
+The order of TreeList columns follows the order of fields in data objects. You can use the [columns](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array to specify a different order. To reorder a column at runtime, change its [visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property.
```
Comment (dirk-pieterse) [inline JSDemos/Demos/TreeList/Reordering/description.md]:
```suggestion
TreeList columns have the same order as fields in data objects. You can use the [columns](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/) array to specify a different order. To reorder a column at runtime, change its [visibleIndex](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visibleIndex) property.
```

**Review by dirk-pieterse — DISMISSED** [review]

**Review by RomanTsukanov — APPROVED** [review]

**Pull request merged**